src/client/failure/policies.ts

```diff
@@ -12,6 +12,71 @@ import {
   SuccessResolution,
 } from './resolution';
 
+export class Router {
+  constructor(
+    private readonly providersOfUseCase: Record<string, FailurePolicy>,
+    private readonly priority: string[],
+    private currentProvider?: string
+  ) {}
+
+  public getCurrentProvider(): string | undefined {
+    return this.currentProvider;
+  }
+
+  public setCurrentProvider(provider: string): void {
+    // TODO: check if exists
+    this.currentProvider = provider;
+  }
+
+  public beforeExecution(info: ExecutionInfo): ExecutionResolution {
+    if (!this.currentProvider) {
+      throw 'Property currentProvider is not set in Router instance';
+    }
+    const innerResolution =
+      this.providersOfUseCase[this.currentProvider].beforeExecution(info);
```

Comment:
At some point we should also attempt going back to previous providers. I think an iteration of the return values of `beforeExecution` here should do the trick.

Reply:
@TheEdward162 I added something very similar and also test for it.